### PR TITLE
Connect LearningResource Drawer to router

### PR DESCRIPTION
**Pull request by ChristopherChudzicki**
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested

#### What are the relevant tickets?
#3734 

#### What's this PR do?
This PR connects the learning resource drawer to the router. Now when you open or close it, the URL updates to reflect that change.

**Additional changes:**
- the `ExpandedLearningResourceDisplay` runs dropdown is now managed wholly within that component
- a some new tests (for the runs dropdown; for the search page; for the drawer routing)
- restructured the info display in `ExpandedLearningResourceDisplay`

#### How should this be manually tested?
1. Open the learning resource drawer on a field page. The URL should update. Close the drawer. The URL should update.
2. Repeat (1) on the search page. In the developer console, check that no extra search API requests are being made when the drawer opens/closes.
3. View the learning resource drawer for courses and videos. Check that it looks correct. _I made some changes to the internal logic of the drawer; there should be essentially no visible difference._
4. Find a course with multiple runs. Check that you can change the run and that the data display updates correctly.

#### Screenshots (if appropriate)

Drawer, **Before** / **After**:

I made some changes to the implementation of the LearningResourceDrawer. There should be basically no visible change. (*Well, "Info" is a little bigger now. And "Offered by" is gray., which seemed like the original intent.)

Screen Shot 2022-09-23 at 11 44 24 AM Screen Shot 2022-09-23 at 12 00 52 PM

### Where to Start?
I would start with `LearningResourceDrawer.tsx` and `RoutedDrawer.tsx` (and its test file).

### CODE REVIEW

File: frontends/infinite-corridor/src/pages/LearningResourceDrawer.tsx [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+          imgConfig={drawerImg}
+          resource={resource}
+          formatShareLink={formatShareLink}
+        />
+      )}
+    </section>
+  )
+}
+
+const LearningResourceDrawer: React.FC = () => {
```
Comment (ChristopherChudzicki) [inline frontends/infinite-corridor/src/pages/LearningResourceDrawer.tsx:58]: @abeglova We had a [slack conversation](https://mitodl.slack.com/archives/D035M0FT7RB/p1661530534282309) about this drawer once:

> I was initially imagining:
> - drawer component in infinite-corridor
> - "LearningResourceDetails" or something in ol-search-ui ... that's the drawer content
> - LearningResourceCards have an onClick prop that we use to set the drawer parameters
> - drawer makes API call and passes relevant info the the details

When I did the work to hook up the drawer to the router, I also implemented that plan.

File: frontends/ol-search-ui/src/components/ExpandedLearningResourceDisplay.test.tsx [hunk 2]
```diff
@@ -1,75 +1,90 @@
...
+    const run = makeRun()
+    const offerer = "ocw"
+    const resource = makeCourseResult({ runs: [run], offered_by: [offerer] })
+
+    const { imgConfig } = renderLearningResourceDetails({ resource })
 
     const instructors = resource.runs[0].instructors
       .map(instructor => getInstructorName(instructor))
       .join(", ")
-    expect(screen.getByText(instructors)).toBeInTheDocument()
```
Comment (ChristopherChudzicki) [inline frontends/ol-search-ui/src/components/ExpandedLearningResourceDisplay.test.tsx:34]: I changed most of the `expect(screen.getBy...).toBeInTheDocument()` to just `screen.getBy...`.

From the [`testing-library`](https://testing-library.com/docs/queries/about/#types-of-queries) docs:
> Returns the matching node for a query, and throw a descriptive error if no elements match or if more than one match is found

File: frontends/ol-search-ui/src/components/ExpandedLearningResourceDisplay.tsx [hunk 3]
```diff
@@ -1,136 +1,119 @@
-import React from "react"
```
Comment (ChristopherChudzicki) [inline frontends/ol-search-ui/src/components/ExpandedLearningResourceDisplay.tsx:1]: Major changes in this file:
- removed `setShowResourceDrawer`... The dropdown management is now local to this component.
- Changed the "info" display at bottom of card to use a [`<dl>` (definition list)](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dl) and extracted the logic for formatting the minformation.

    `<dl>`s aren't too common, but I think it is a good fit here.

File: frontends/ol-util/src/predicates.ts [hunk 4]
```diff
@@ -24,6 +24,26 @@ export const isNotNil = <T>(x: T): x is NonNullable<T> => {
...
+ * the returned function `obj => boolean` is a Typescript Predicate that can
+ * be used to filter arrays. For example:
+ *
+ * ```ts
+ * type A = { a: number }
+ * const maybes = [{ a: 1 }, {}, { a: undefined }]
+ * const definitelies: A[] = maybes.filter(propsNotNil(["a"]))
+ * ```
+ */
+export const propsNotNil = <P extends string>(propNames: P[]) => {
```
Comment (ChristopherChudzicki) [inline frontends/ol-util/src/predicates.ts:39]: Typescript 4.7 did not like this, but Typescript 4.8 did, which is why I upgraded.

_And how I discovered that my VSCode was using 4.8 not the version specified in our package.json...._

**Review by abeglova — APPROVED** [review]
LGTM

**Pull request merged**